FLIGHT_DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,442 @@
+    collision, we know that it was this flight that previously wrote the row and we can proceed to the next step.
+    If all concurrent creators check the database first, they will conflict on the name
+    before doing any creating.
+ 1. Create the target object
+ 1. If needed, update the database row to record details of the created object.
+
+These pitfalls are presented in the context of object creation, but they appear in related
+forms in other kinds of operations.
+
+## Retrying Steps in a Flight <a name="retrying"></a>
+
+### Retry Object State and Reuse <a name="retryreuse"></a>
+It is important to understand the scope and state of a RetryRule object and only share
+RetryRules objects where they will work properly. It is safe to re-use the same RetryRule
+object *within* a flight. It is not safe to share the same RetryRule object *between*
+flights.
+
+The reason that you cannot share a RetryRule object between flights is that a RetryRule
+object is stateful: it holds the state of retrying. For example, a rule that retries 3
+times needs to hold a counter for the number of retries that have been done. If that
+object is used across flights, each flight would be updating the number of retries.
+
+The reason that you can share a RetryRule object within a flight is that RetryRules have
+an `initialize()` method that resets the state. The Flight object always initializes the
+RetryRule object at the start of executing the `do()` and `undo()` methods of a Step.
+
+Every step has a RetryRule object associated with it. If the rule is not specified, the
+`RetryRuleNone` is used. RetryRuleNone has no state and does no retries. It always
+returns: "do not retry".
+
+### How to Retry <a name="retryhow"></a>
+There are two ways from with a step to request that the step be retried.
+ 1. Throw an exception that derives from `RetryException`. When flight execution catches
+ that exception, it will attempt a retry.
+ 1. Return a StepResult of `STEP_RESULT_FAILURE_RETRY`.
+
+It is our experience that having the step perform its own exception handling and return
+the explicit `STEP_RESULT_FAILURE_RETRY` is more readable. It also keeps service
+exceptions separate from Stairway exceptions.
+
+Remember that simply requesting a retry and actually retrying are two separate
+things. Requesting retry asks Stairway to check the RetryRule to see if there are any more
+retries left. If the rule says, "no more retries", then the exception returned in the
+StepResult is used as the result exception of the flight.
+
+### Retry After Restart <a name="retryafterrestart"></a>
+Retrying is only in-memory. RetryRule state is not retained across a failure/recovery.
+
+Imagine that Stairway is running a step of a flight that uses a retry rule. The rule will
+try 10 times and then give up. Stairway has retried 9 times and then the server fails.
+When the flight is recovered, the step will be restarted and the retry rule counter will
```

Comment:
Is there a limit to how many times a flight can be recovered?

Reply:
No, there is no limit.